nomos-da/network/core/src/protocols/replication/behaviour.rs

```diff
@@ -156,361 +218,254 @@ where
-
-        behaviours[i]
-            .handle_established_outbound_connection(
-                connection_id,
-                peer_id_j,
-                &Multiaddr::empty(),
-                Endpoint::Dialer,
-                Default::default(),
-            )
-            .unwrap();
-
-        behaviours[j]
-            .handle_established_inbound_connection(
-                connection_id,
-                peer_id_i,
-                &Multiaddr::empty(),
-                &Multiaddr::empty(),
-            )
-            .unwrap();
-    }
-
-    fn deliver_message_to_peer(
-        all_behaviours: &mut [ReplicationBehaviour<MockMembershipHandler>],
-        peer_id: PeerId,
-        connection_id: ConnectionId,
-        message: DaMessage,
-    ) {
-        if let Some(behaviour) = all_behaviours
-            .iter_mut()
-            .find(|b| b.local_peer_id == peer_id)
-        {
-            // Simulate the handler receiving the message.
-            behaviour.on_connection_handler_event(
-                peer_id,
-                connection_id,
-                Either::Left(HandlerEventToBehaviour::IncomingMessage { message }),
-            );
-        }
-    }
-
-    #[tokio::test]
-    async fn test_replication_behaviour() {
-        let num_instances = 20;
-        let mut membership = HashMap::default();
-
-        let subnet_0_behaviours =
-            create_replication_behaviours(num_instances / 2, 0, &mut membership);
-        let subnet_1_behaviours =
-            create_replication_behaviours(num_instances / 2, 1, &mut membership);
-
-        let mut all_behaviours = subnet_0_behaviours;
-        all_behaviours.extend(subnet_1_behaviours);
-
-        for behaviour in all_behaviours.iter_mut() {
-            let membership_handler = MockMembershipHandler {
-                membership: membership.clone(),
-            };
-            behaviour.update_membership(membership_handler);
-        }
-
-        // Simulate peer connections.
-        for (i, j) in (0..num_instances).flat_map(|i| (i + 1..num_instances).map(move |j| (i, j))) {
-            let connection_id = ConnectionId::new_unchecked(i);
-            establish_connection(&mut all_behaviours, i, j, connection_id);
+                    // No more messages to send, put the stream back in the idle state
+                    None => {
+                        self.outgoing_streams
+                            .insert(peer_id, StreamState::Idle(stream));
+                    }
+                };
+            }
+            Poll::Ready(Some(Err(error))) => {
+                debug!("Error on replication stream {error:?}");
+                // TODO propagate error ?
+            }
+            _ => {}
         }
+        // Schedule writing pending messages to connected peers if possible,
+        // ie. if the streams for those peers are idle or haven't even been opened yet
+        for (peer_id, messages) in &mut self.pending_outgoing_messages {
```

Comment:
Lets also process one `pendiding_outgoing_message` at a time, just in case we have a lot and would like other behaviours to make some progress.